src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java

```diff
@@ -79,17 +112,141 @@ enum Signal {
+    }
+
+    response = sendRequest(new Request.GetEnvironment());
+    if (response instanceof Response.Environment environmentResponse) {
+      return environmentResponse.environment;
+    } else {
+      throw new IllegalStateException("Unexpected response %s" + response);
+    }
+  }
+
+  private void logMaybe(String msg) {
+    // System.err.println(msg);
   }
 
+  Response sendRequest(Request request) throws InterruptedException {
+    int seq = nextRequestSequenceNo++;
+    Packet<Request> packet = new Packet(seq, request);
+    logMaybe(String.format("LOG %s/worker: sending request %s", repositoryName, packet));
+    requestQueue.put(packet);
+    while (true) {
+      Packet<Response> response = responseQueue.take();
+      if (response.sequenceNo == seq) {
+        return response.message;
+      } else {
+        logMaybe(String.format("LOG %s/worker: discarding %s", repositoryName, response));
+      }
+    }
+  }
+
+  Response sendRequestUninterruptibly(Request request) {
+    int seq = nextRequestSequenceNo++;
+    Packet<Request> packet = new Packet(seq, request);
+    logMaybe(String.format("LOG %s/worker: sending request uninterruptibly %s", repositoryName, packet));
+    Uninterruptibles.putUninterruptibly(requestQueue, packet);
+    while (true) {
+      Packet<Response> response = Uninterruptibles.takeUninterruptibly(responseQueue);
+      logMaybe(String.format("LOG %s/worker: got response packet %s", repositoryName, response));
+      if (response.sequenceNo == seq) {
+        return response.message;
+      }
+    }
+  }
+
+  Packet<Request> getRequest() throws InterruptedException {
+    logMaybe(String.format("LOG %s/host: taking request", repositoryName));
+    Packet<Request> result = requestQueue.take();
+    logMaybe(String.format("LOG %s/host: Got request %s", repositoryName, result));
+    return result;
+  }
+
+  Packet<Request> getRequestUninterruptibly() {
+    logMaybe(String.format("LOG %s/host: taking request uninterruptibly", repositoryName));
+    Packet<Request> result = Uninterruptibles.takeUninterruptibly(requestQueue);
+    logMaybe(String.format("LOG %s/host: request received %s", repositoryName, result));
+    return result;
+  }
+
+  void sendResponse(Packet<Request> request, Response response) throws InterruptedException {
+    logMaybe(String.format("LOG %s/host: sending response %s", repositoryName, response));
+    Packet<Response> packet = new Packet<>(request.sequenceNo, response);
+    responseQueue.put(packet);
+  }
+
+  void sendResponseUninterruptibly(Packet<Request> request, Response response) {
+    logMaybe(String.format("LOG %s/host: sending response uninterruptibly %s", repositoryName, response));
+    Packet<Response> packet = new Packet<>(request.sequenceNo, response);
+    Uninterruptibles.putUninterruptibly(responseQueue, packet);
+  }
+
+  /** Join the worker thread.
+   *
+   * This method should only bee called if it either sent a {@code Failure} message or it received
```

Comment:
nit: insert checks for the preconditions?

Reply:
I tried! Turns out, it's not that easy because this condition means that the "finished" bit must be set on the worker thread, but can only be set in the success case after it receives the answer not to restart, but then it races with the host thread's `join()` call.

I could get around that by adding a new message that unambiguously indicates the termination of the worker thread, but that sounds like it'd cost more than what it's worth.